src/fs/rock/RockRebuild.cc

```diff
@@ -156,92 +203,138 @@ class LoadingParts
+    flags(shm_new(Flags)(FlagsPath(dir->path).c_str(), dir->slotLimitActual()))
+{
+    auto moresArray = mores->object();
+    for (int i = 0; i < moresArray->capacity; ++i)
+        moresArray->items[i] = -1;
+}
+
+Rock::LoadingPartsOwner::~LoadingPartsOwner()
+{
+    delete sizes;
+    delete versions;
+    delete mores;
+    delete flags;
+}
+
 /* LoadingEntry */
 
 Rock::LoadingEntry::LoadingEntry(const sfileno fileNo, LoadingParts &source):
-    size(source.sizes.at(fileNo)),
-    version(source.versions.at(fileNo)),
-    flags(source.flags.at(fileNo))
+    size(source.sizes->items[fileNo]),
+    version(source.versions->items[fileNo]),
+    flags(source.flags->items[fileNo])
 {
 }
 
 /* LoadingSlot */
 
 Rock::LoadingSlot::LoadingSlot(const SlotId slotId, LoadingParts &source):
-    more(source.mores.at(slotId)),
-    flags(source.flags.at(slotId))
+    more(source.mores->items[slotId]),
+    flags(source.flags->items[slotId])
 {
 }
 
 /* LoadingParts */
 
-Rock::LoadingParts::LoadingParts(const int dbEntryLimit, const int dbSlotLimit):
-    sizes(dbEntryLimit, 0),
-    versions(dbEntryLimit, 0),
-    mores(dbSlotLimit, -1),
-    flags(dbSlotLimit)
+Rock::LoadingParts::LoadingParts(const char *dirPath):
+    sizes(shm_old(Sizes)(SizesPath(dirPath).c_str())),
+    versions(shm_old(Versions)(VersionsPath(dirPath).c_str())),
+    mores(shm_old(Mores)(MoresPath(dirPath).c_str())),
+    flags(shm_old(Flags)(FlagsPath(dirPath).c_str()))
 {
-    assert(sizes.size() == versions.size()); // every entry has both fields
-    assert(sizes.size() <= mores.size()); // every entry needs slot(s)
-    assert(mores.size() == flags.size()); // every slot needs a set of flags
+    assert(sizes->capacity == versions->capacity); // every entry has both fields
+    assert(sizes->capacity <= mores->capacity); // every entry needs slot(s)
+    assert(mores->capacity == flags->capacity); // every slot needs a set of flags
 }
 
 /* Rebuild */
 
 Rock::Rebuild::Rebuild(SwapDir *dir): AsyncJob("Rock::Rebuild"),
     sd(dir),
     parts(nullptr),
+    metadata(shm_old(Metadata)(MetadataPath(dir->path).c_str())),
     dbSize(0),
     dbSlotSize(0),
     dbSlotLimit(0),
     dbEntryLimit(0),
     fd(-1),
     dbOffset(0),
-    loadingPos(0),
-    validationPos(0)
+    loadingPos(metadata->counts.scancount),
+    validationPos(metadata->counts.validatedCount),
+    counts(metadata->counts),
+    partsOwner(nullptr),
+    resuming(metadata->counts.started())
 {
     assert(sd);
     dbSize = sd->diskOffsetLimit(); // we do not care about the trailer waste
     dbSlotSize = sd->slotSize;
     dbEntryLimit = sd->entryLimitActual();
     dbSlotLimit = sd->slotLimitActual();
     assert(dbEntryLimit <= dbSlotLimit);
+    registerRunner();
 }
 
 Rock::Rebuild::~Rebuild()
 {
     if (fd >= 0)
         file_close(fd);
     delete parts;
+    delete partsOwner;
+    if (resuming && loadedAndValidated()) {
+        Ipc::Mem::Segment::Unlink(SizesPath(sd->path).c_str());
```

Comment:
Let's clarify this unusual code a little:

```suggestion
        // normally, segments are used until the Squid instance quits,
        // but these indexing-only segments are no longer needed
        Ipc::Mem::Segment::Unlink(SizesPath(sd->path).c_str());
```